superset/tables/models.py

```diff
@@ -80,13 +107,96 @@ class Table(Model, AuditMixinNullable, ExtraJSONMixin, ImportExportMixin):
     schema = sa.Column(sa.Text)
     name = sa.Column(sa.Text)
 
-    # The relationship between tables and columns is 1:n, but we use a many-to-many
-    # association to differentiate between the relationship between datasets and
-    # columns.
-    columns: List[Column] = relationship(
-        "Column", secondary=association_table, cascade="all, delete"
-    )
-
     # Column is managed externally and should be read-only inside Superset
     is_managed_externally = sa.Column(sa.Boolean, nullable=False, default=False)
     external_url = sa.Column(sa.Text, nullable=True)
+
+    @property
+    def fullname(self) -> str:
+        return str(TableName(table=self.name, schema=self.schema, catalog=self.catalog))
+
+    def __repr__(self) -> str:
+        return f"<Table id={self.id} database_id={self.database_id} {self.fullname}>"
+
+    def sync_columns(self) -> None:
+        """Sync table columns with the database. Keep metadata for existing columns"""
+        try:
+            column_metadata = get_physical_table_metadata(
+                self.database, self.name, self.schema
+            )
+        except Exception:  # pylint: disable=broad-except
+            column_metadata = []
+
+        existing_columns = {column.name: column for column in self.columns}
+        quote_identifier = self.database.quote_identifier
+
+        def update_or_create_column(column_meta: Dict[str, Any]) -> Column:
+            column_name: str = column_meta["name"]
+            if column_name in existing_columns:
+                column = existing_columns[column_name]
+            else:
+                column = Column(name=column_name)
+            column.type = column_meta["type"]
+            column.is_temporal = column_meta["is_dttm"]
+            column.expression = quote_identifier(column_name)
+            column.is_aggregation = False
+            column.is_physical = True
+            column.is_spatial = False
+            column.is_partition = False  # TODO: update with accurate is_partition
+            return column
+
+        self.columns = list(map(update_or_create_column, column_metadata))
+
+    @staticmethod
+    def load_or_create(
+        database: Database,
+        table_names: Iterable[TableName],
+        default_schema: Optional[str] = None,
+        sync_columns: Optional[bool] = False,
+        default_props: Optional[Dict[str, Any]] = None,
+    ) -> List["Table"]:
+        """
+        Load or create new table model instances.
+        """
+        if not table_names:
+            return []
+
+        if not database.id:
+            raise Exception("Database must be already committed")
```

Comment:
You get an `id` by calling `session.flush()` as well, I think that's still fine, right?

```suggestion
            raise Exception("Database must be already flushed or committed")
```